### PR TITLE
`expr`: short-circuit evaluation for `|`

### LINKED ISSUE #5326: ` expr 1 \| a / 5` is producing different results than GNU
With GNU:

```
 /usr/bin/expr 86 \> 1 \|  aR / 5
 echo $?
1
0
```

With our implementation:
```
./target/debug/coreutils expr 86 \> 1 \|  a / 5 
echo $?
expr: Expected an integer operand
2
```

### PR DESCRIPTION
**Pull request by Luv-Ray**
fix #5326 
Implement short-circuit evaluation for `|`. And add the corresponding test.

### CODE REVIEW

File: src/uu/expr/src/syntax_tree.rs [hunk 1]
```diff
@@ -155,8 +155,24 @@ impl AstNode {
...
-        if let Self::Node { operands, .. } = self {
+        if let Self::Node {
+            operands, op_type, ..
+        } = self
+        {
             let mut out = Vec::with_capacity(operands.len());
+            let mut operands = operands.iter();
+
+            if op_type == "|" {
+                if let Some(value) = operands.next() {
```
Comment (sylvestre) [inline src/uu/expr/src/syntax_tree.rs:166]: could you please add some comments to describe this ? 

and explain what it is doing :)

thanks

Comment (Luv-Ray) [reply src/uu/expr/src/syntax_tree.rs:167]: I have add a comment to describe it :)

File: tests/by-util/test_expr.rs [hunk 2]
```diff
@@ -123,6 +123,11 @@ fn test_or() {
         .args(&["-14", "|", "1"])
         .succeeds()
         .stdout_only("-14\n");
+
```
Comment (sylvestre) [inline tests/by-util/test_expr.rs:126]: could you please add more tests to test this behavior ? :)
thanks

Comment (Luv-Ray) [reply tests/by-util/test_expr.rs:126]: Add two more tests, and they can be passed on my local machine :)

File: tests/by-util/test_expr.rs [hunk 3]
```diff
@@ -123,6 +123,21 @@ fn test_or() {
...
+    
+    new_ucmd!()
+        .args(&["foo", "|", "a", "/", "5"])
+        .succeeds()
+        .stdout_only("foo\n");
+
+    new_ucmd!()
+        .args(&["0", "|", "10", "/", "5"])
+        .succeeds()
+        .stdout_only("2\n");
```
Comment (sylvestre) [inline tests/by-util/test_expr.rs:140]: do you know why the result is 2 here?

Comment (Luv-Ray) [reply tests/by-util/test_expr.rs:140]: if the first value is 0, which means `value_as_bool(&value) == false`,
then the code i have modified will executed as below:
```rust
if op_type == "|" {
    if let Some(value) = operands.next() {
        let value = value.evaluate()?;
        out.push(value.clone());
    }
}
```
just do the same thing as origin code (`next` do the same thing as `for`), and will not effect the evaluate of `10 / 5`.
GNU produce the same result:
```shell
$ expr 0 \| 10 / 5
2
```

**Comment by Luv-Ray** [comment]
As I am not familiar with PR, I may have done something wrong, sorry about that. I think the branch is clean now.
Is the implementation in the code inappropriate or incorrect?

File: tests/by-util/test_expr.rs (hunk 3 above)
Comment (sylvestre) [reply tests/by-util/test_expr.rs:140]: ok, thanks :)

**Comment by cakebaker** [comment]
Thanks for your PR :)

**Pull request merged**